File: MainApp/src/main/java/com/sleticalboy/dailywork/weight/xrecycler/adapter/XBaseHolder.java
package com.sleticalboy.dailywork.weight.xrecycler.adapter;

import android.content.Context;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.reflect.Field;

/**
 * Created on 18-2-7.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
public abstract class XBaseHolder<M> extends RecyclerView.ViewHolder {

    public XBaseHolder(View itemView) {
        super(itemView);
    }

    public XBaseHolder(ViewGroup parent, @LayoutRes int res) {
        this(LayoutInflater.from(parent.getContext()).inflate(res, parent, false));
    }

    protected <V extends View> V getView(@IdRes int id) {
        return (V) itemView.findViewById(id);
    }

    protected int getDataPosition() {
        RecyclerView.Adapter adapter = getOwnerAdapter();
        if (adapter != null && adapter instanceof XRecyclerAdapter) {
            return getAdapterPosition() - ((XRecyclerAdapter) adapter).getHeadersCount();
        }
        return getAdapterPosition();
    }

    @Nullable
    protected <A extends RecyclerView.Adapter> A getOwnerAdapter() {
        RecyclerView recyclerView = getOwnerRecyclerView();
        return recyclerView == null ? null : (A) recyclerView.getAdapter();
    }

    protected RecyclerView getOwnerRecyclerView() {
        try {
            Field mOwnerRecyclerView = RecyclerView.ViewHolder.class.getDeclaredField("mOwnerRecyclerView");
            mOwnerRecyclerView.setAccessible(true);
            return (RecyclerView) mOwnerRecyclerView.get(this);
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        return null;
    }

    protected Context getContext() {
        return itemView.getContext();
    }

    protected abstract void setData(M data);
}

File: MainApp/src/main/java/com/sleticalboy/weight/xrecycler/adapter/XBaseHolder.kt
package com.sleticalboy.weight.xrecycler.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView

/**
 * Created on 18-2-7.
 *
 * @author leebin
 * @version 1.0
 * @description
 */
abstract class XBaseHolder<M>(itemView: View?) : RecyclerView.ViewHolder(itemView!!) {

    constructor(parent: ViewGroup, @LayoutRes res: Int)
            : this(LayoutInflater.from(parent.context).inflate(res, parent, false))

    protected fun <V : View?> getView(@IdRes id: Int): V {
        return itemView.findViewById<View>(id) as V
    }

    protected val dataPosition: Int
        get() {
            val adapter = getOwnerAdapter<RecyclerView.Adapter<*>>()
            return if (adapter != null && adapter is XRecyclerAdapter<*>) {
                adapterPosition - adapter.headersCount
            } else adapterPosition
        }

    protected fun <A : RecyclerView.Adapter<*>?> getOwnerAdapter(): A? {
        val recyclerView = ownerRecyclerView
        return if (recyclerView == null) null else recyclerView.adapter as A?
    }

    private val ownerRecyclerView: RecyclerView?
        get() {
            try {
                val mOwnerRecyclerView = RecyclerView.ViewHolder::class.java
                        .getDeclaredField("mOwnerRecyclerView")
                mOwnerRecyclerView.isAccessible = true
                return mOwnerRecyclerView[this] as RecyclerView
            } catch (e: NoSuchFieldException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
            return null
        }

    protected val context: Context get() = itemView.context

    abstract fun bindData(data: M)
}